### PR TITLE
feat(APIM-90-91): post and get premium schedules

### PR DESCRIPTION
**Pull request by avaitonis**
### Introduction
New endpoint POST /premium/schedule . 
New endpoint GET /premium/segments/{facilityId} . Get previously generated premium schedules.

###  Resolution
POST /premium/schedule 

- uses SP - USP_MDM_INCOME_EXPOSURE
- generates and returns premium schedules
- returns Location header for GET call

GET  /premium/segments/{facilityId} 

- gets data from table using typeOrm

Allow referencing third party module "express" in .eslintrc.json
Unit test has simple Response mock/fake/stub. It could be improved in future.

### CODE REVIEW

File: src/modules/premium-schedules/premium-schedules.service.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+      res.set({ Location: '/premium/segments/' + createPremiumSchedule.facilityURN });
+
+      // Older Mulesoft implementation return Location header and empty payload, but here we already have new segments and can return them.
+      const fieldMap = DbResponseHelper.getApiNameToDbNameMap(this.premiumSchedulesRepository);
+
+      const renamedResults = DbResponseHelper.renameDbResultFields(this.premiumSchedulesRepository, fieldMap, spResults);
+
+      // Transform results to match logic in old implementation.
+      // Remove time part of field calculationDate.
+      renamedResults.forEach((result) => (result.calculationDate = result.calculationDate.toISOString().split('T')[0]));
```
Comment (IgnatG) [inline src/modules/premium-schedules/premium-schedules.service.ts:84]: can we use `.map` here instead? i.e. `renamedResults.map({..})`

File: test/premium-schedules/premium-schedules.api-test.ts [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+import Chance from 'chance';
+
+import { Api } from '../api';
+import { CreateApp } from '../createApp';
+
+const chance = new Chance();
+
+describe('Premium schedules', () => {
+  let app: INestApplication;
+  let api;
```
Comment (IgnatG) [inline test/premium-schedules/premium-schedules.api-test.ts:11]: we can add the type to this as well `let api: Api;`

**Review by IgnatG — APPROVED** [review]
LGTM - I left 2 comments, but they are small and can be addressed as part of a separate PR if needed

File: src/modules/premium-schedules/dto/create-premium-schedule.dto.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
+import { ApiProperty } from '@nestjs/swagger';
+import { IsDateString, IsInt, IsNotEmpty, IsNumber, IsOptional, IsString, Length, MaxLength } from 'class-validator';
+
+export class CreatePremiumScheduleDto {
+  @IsInt()
+  @IsNotEmpty()
+  @ApiProperty({ example: 10588388, description: 'Can be BS or EW' })
```
Comment (lmarrai-sw) [inline src/modules/premium-schedules/dto/create-premium-schedule.dto.ts:7]: What are BS and EW?

File: src/modules/premium-schedules/dto/create-premium-schedule.dto.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  @IsNotEmpty()
+  @Length(2, 2)
+  @ApiProperty({ example: 'BS', description: 'Can be BS or EW' })
+  readonly productGroup: string;
+
+  @IsNumber()
+  @IsNotEmpty()
+  @ApiProperty({
+    example: 1,
+    description: 'Premium type concerns how we are being paid... In Arrears, In advance or one lump sum. See MASTER_DATA.dbo.WF_PREMIUM_TYPE',
```
Comment (lmarrai-sw) [inline src/modules/premium-schedules/dto/create-premium-schedule.dto.ts:20]: @IgnatG mentioned that we might want to allow teams across UKEF to use this documentation - will `MASTER_DATA.dbo.WF_PREMIUM_TYPE` mean much to them?

File: src/modules/premium-schedules/dto/create-premium-schedule.dto.ts (hunk 3 above)
Comment (IgnatG) [reply src/modules/premium-schedules/dto/create-premium-schedule.dto.ts:7]: different types of deal (Bond Support and Export Working)

File: src/modules/premium-schedules/dto/create-premium-schedule.dto.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  @IsInt()
+  @IsNotEmpty()
+  @ApiProperty({ example: 10588388, description: 'Can be BS or EW' })
+  readonly facilityURN: number;
+
+  @IsString()
+  @IsNotEmpty()
+  @Length(2, 2)
+  @ApiProperty({ example: 'BS', description: 'Can be BS or EW' })
+  readonly productGroup: string;
```
Comment (lmarrai-sw) [inline src/modules/premium-schedules/dto/create-premium-schedule.dto.ts:14]: If this can only be `BS` or `EW`  then should we validate this?

File: src/modules/premium-schedules/dto/create-premium-schedule.dto.ts (hunk 3 above)
Comment (lmarrai-sw) [reply src/modules/premium-schedules/dto/create-premium-schedule.dto.ts:7]: thanks @IgnatG!

File: src/modules/premium-schedules/dto/get-premium-schedule-param.dto.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import { ApiProperty } from '@nestjs/swagger';
+import { IsNotEmpty, IsString, Matches } from 'class-validator';
+
+export class GetPremiumScheduleParamDto {
+  @IsString()
+  @IsNotEmpty()
+  @ApiProperty({ example: '10588388' })
```
Comment (lmarrai-sw) [inline src/modules/premium-schedules/dto/get-premium-schedule-param.dto.ts:7]: Should we have a description as well?

File: src/modules/premium-schedules/dto/create-premium-schedule.dto.ts (hunk 5 above)
Comment (avaitonis) [reply src/modules/premium-schedules/dto/create-premium-schedule.dto.ts:14]: Do we have example how to create custom validator?
For now I will add regex as in src\modules\exposure-period\dto\get-exposure-period-query.dto.ts

File: src/modules/premium-schedules/premium-schedules.service.ts (hunk 1 above)
Comment (avaitonis) [reply src/modules/premium-schedules/premium-schedules.service.ts]: Changed.

File: test/premium-schedules/premium-schedules.api-test.ts (hunk 2 above)
Comment (avaitonis) [reply test/premium-schedules/premium-schedules.api-test.ts]: Changed.

File: src/modules/premium-schedules/dto/create-premium-schedule.dto.ts (hunk 3 above)
Comment (avaitonis) [reply src/modules/premium-schedules/dto/create-premium-schedule.dto.ts]: Description was wrong. Updated.

In this project and TFS-API BS (aka BSS) and EW (aka EWCS) are type of Facilities.

File: src/modules/premium-schedules/dto/create-premium-schedule.dto.ts (hunk 4 above)
Comment (avaitonis) [reply src/modules/premium-schedules/dto/create-premium-schedule.dto.ts]: I checked table and very few premium types are available. Added possible value list to the description.

File: src/modules/premium-schedules/dto/get-premium-schedule-param.dto.ts (hunk 6 above)
Comment (avaitonis) [reply src/modules/premium-schedules/dto/get-premium-schedule-param.dto.ts]: Added

**Review by IgnatG — APPROVED** [review]

File: src/modules/premium-schedules/premium-schedules.service.ts [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+      // Older Mulesoft implementation return Location header and empty payload, but here we already have new segments and can return them.
+      const fieldMap = DbResponseHelper.getApiNameToDbNameMap(this.premiumSchedulesRepository);
+
+      const renamedResults = DbResponseHelper.renameDbResultFields(this.premiumSchedulesRepository, fieldMap, spResults);
+
+      // Transform results to match logic in old implementation.
+      const transformedResults = renamedResults.map((result) => {
+        // Remove time part of Date field calculationDate.
+        result.calculationDate = result.calculationDate.toISOString().split('T')[0];
```
Comment (abhi-markan) [inline src/modules/premium-schedules/premium-schedules.service.ts:85]: Can we please create a helper function where timebased splitting is done please?
Happy for this to be a separate PR.

**Review by abhi-markan — APPROVED** [review]
Looks one minor comment (not a blocker, can be a future PR).

File: src/modules/premium-schedules/premium-schedules.service.ts (hunk 7 above)
Comment (avaitonis) [reply src/modules/premium-schedules/premium-schedules.service.ts:85]: Will do in future, especially if it will be required elsewhere
